README.md

```diff
@@ -1,6 +1,101 @@
+# Introduction #
+
+## The problem ##
+
+The microservices architecture introduces many new challenges when dealing with complex systems. One of them is the high number of remote procedure calls (RPCs) and the cost associated to them. Among the common techniques applied to amortize this cost, batching of requests has an important role. Instead of paying the price of one call for each interaction, many interactions are batched in only one call.
+
+While batching introduces performance enhancements, it also introduces complexity to the codebase. The common approach is to extract as much information as possible about what needs to be fetched, perform the batched fetch, and then compose the values again to have the final result. These steps need to repeated many times depending on how complex is the final structure.
+
+An example of batching using futures:
+
+```scala
+tracksService.get(ids).flatMap { tracks =>
+	val userIds = tracks.map(_.user)
+	usersService.get(userIds).map {	users =>
+		val userMap = userIds.zip(users).toMap
+		tracks.map { track =>
+			new EnrichedTrack(track, userMap(track.user))
+		}
+	}
+}
+```
+
+This example has only one level of nested resources. In a complex system, it is common to have several levels:
+
+```
+timeline
+	track post
+		track
+			creator
+	track repost
+		track
+			creator
+		reposter
+	playlist post
+		playlist
+			track ids
+			creator
+	playlist repost
+		playlist
+			track ids
+			creator
+		reposter
+	comment
+		user
+	user follow
+		follower
+		followee
+```
+
+This structure can also be part of a bigger structure that includes the user's data for instance. Considering the corner cases where the fetch result can be optional, the final code that is capable of batching requests in an optimal way is really complex and hard to maintain.
+
+## The solution ##
+
+The complexity comes mainly from declaring together **what** needs to be fetched and **how** it should be fetched. Clump offers an embedded Domain Specific Language (DSL) that allows to declare **what** needs to be fetched and an execution model that determine **how** the resources should be fetched. This execution model has some optimizations to perform requests concurrently, use batches when possible and avoid fetching the same resource multiple times by using a cache.
+
+The DSL is implemented using a monadic interface similar to ```Future```. It is a Free Monad, that only produces a nested series of transformations without triggering the actual execution. This is the characteristic that allows to build the execution model separately from the definition of what needs to be fetched.
+
+The execution model leverages on Applicative Functors, that allows to express independence of computations. It exposes only ```join``` to the user but makes heavy use of other applicative operations internally. This means that even without the user specifying what is independent, the execution model is capable to apply the optimizations.
+
```

Comment:
It is necessary to find good sources for these functional concepts and add links.